sycl/test-e2e/Matrix/slm_utils.hpp

```diff
@@ -0,0 +1,163 @@
+//==------------------ slm_utils.hpp  - DPC++ joint_matrix------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#ifdef OCL
+typedef ushort ushort16ocl __attribute__((ext_vector_type(16)));
+typedef ushort ushort8ocl __attribute__((ext_vector_type(8)));
+typedef ushort ushort4ocl __attribute__((ext_vector_type(4)));
+typedef ushort ushort2ocl __attribute__((ext_vector_type(2)));
+
+extern "C" {
+SYCL_EXTERNAL ushort2ocl __builtin_IB_simd_block_read_2_global_h(
+    __attribute__((opencl_global)) ushort *base);
+SYCL_EXTERNAL void __builtin_IB_simd_block_write_2_local_h(
+    __attribute__((opencl_local)) ushort *base, ushort2ocl);
+SYCL_EXTERNAL ushort4ocl __builtin_IB_simd_block_read_4_global_h(
+    __attribute__((opencl_global)) ushort *base);
+SYCL_EXTERNAL void __builtin_IB_simd_block_write_4_local_h(
+    __attribute__((opencl_local)) ushort *base, ushort4ocl);
+SYCL_EXTERNAL ushort8ocl __builtin_IB_simd_block_read_8_global_h(
+    __attribute__((opencl_global)) ushort *base);
+SYCL_EXTERNAL void __builtin_IB_simd_block_write_8_local_h(
+    __attribute__((opencl_local)) ushort *base, ushort8ocl);
+SYCL_EXTERNAL ushort16ocl __builtin_IB_simd_block_read_16_global_h(
+    __attribute__((opencl_global)) ushort *base);
+SYCL_EXTERNAL void __builtin_IB_simd_block_write_16_local_h(
+    __attribute__((opencl_local)) ushort *base, ushort16ocl);
+};
+
+template <unsigned int rowsA, unsigned int colsA, unsigned int rowsB,
+          unsigned int colsB, unsigned int MCACHE2, unsigned int NCACHE2,
+          unsigned int KCACHE2, unsigned int vnniFactor, size_t SGs,
+          typename TOperand>
+inline void
+slm_read_write_OCL(TOperand *A, TOperand *B, local_accessor<TOperand, 2> tileA,
+                   local_accessor<TOperand, 2> tileB, sub_group sg,
+                   unsigned int k2, size_t m2, size_t n2, size_t SG_SIZE) {
+
+  size_t slmReadA = KCACHE2 / SG_SIZE;
+
+  for (int i = 0; i < (MCACHE2 / SGs); i++) {
+    __attribute__((opencl_global)) ushort *g_addrA;
+    g_addrA =
+        (__attribute__((opencl_global))
+         ushort *)(A +
+                   (m2 * MCACHE2 + (sg.get_group_id()) * (MCACHE2 / SGs) + i) *
+                       colsA +
+                   k2 * KCACHE2);
+    __attribute__((opencl_local)) ushort *l_addrA;
+    l_addrA =
+        (__attribute__((opencl_local)) ushort
+             *)(tileA.template get_multi_ptr<sycl::access::decorated::yes>() +
```

Comment:
For my understanding, why do we need decorated::yes in OCL SLM calls, at the same time in SYCL in is sg.store, decorated::no is used?

Reply:
Actually, the local ones can have `decorated::no`. I will change it because `decorated::yes` is not needed there.